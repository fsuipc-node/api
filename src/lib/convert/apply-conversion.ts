import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
