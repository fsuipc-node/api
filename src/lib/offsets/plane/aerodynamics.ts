import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  incidenceAlpha: new Offset({
    value: 0x2ED0,
    name: 'incidenceAlpha',
    description: 'incidence alpha (body AoA) - radians',
    category: OffsetCategory.AERODYNAMICS,
    type: Type.Double,
    permission: 'r',
  }),
};
