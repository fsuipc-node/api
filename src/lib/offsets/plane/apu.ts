import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  APUGeneratorSwitch: new Offset({
    value: 0xB51,
    name: 'APUGeneratorSwitch',
    category: OffsetCategory.APU,
    description: 'APU generator switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  APUGeneratorActive: new Offset({
    value: 0xB52,
    name: 'APUGeneratorActive',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUOnFire: new Offset({
    value: 0xB53,
    name: 'APUOnFire',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUMaxRPM: new Offset({
    value: 0xB54,
    name: 'APUMaxRPM',
    category: OffsetCategory.APU,
    description: 'APU generator max RPM percent',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'r',
  }),
  APUStarter: new Offset({
    value: 0xB58,
    name: 'APUStarter',
    category: OffsetCategory.APU,
    description: 'APU starter',
    convert: '{VAL} ? 1 : 0',
    type: Type.Single,
    permission: 'rw',
  }),
  APUVoltage: new Offset({
    value: 0xB5C,
    name: 'APUVoltage',
    category: OffsetCategory.APU,
    description: 'APU generator voltage',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'rw',
  }),
};
