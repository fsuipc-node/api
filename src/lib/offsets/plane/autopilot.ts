import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const autopilot: OffsetList = {
  autoPilotAvailable: new Offset({
    value: 0x764,
    name: 'autoPilotAvailable',
    category: OffsetCategory.AUTOPILOT,
    description: 'autopilot available',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  flyByWireELACSwitch: new Offset({
    value: 0x7B6,
    name: 'flyByWireELACSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire ELAC switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  flyByWireELACCompFailFlag: new Offset({
    value: 0x7B7,
    name: 'flyByWireELACCompFailFlag',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire ELAC comp fail flag (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  flyByWireFACSwitch: new Offset({
    value: 0x7B8,
    name: 'flyByWireFACSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire FAC switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  flyByWireFACCompFailFlag: new Offset({
    value: 0x7B9,
    name: 'flyByWireFACCompFailFlag',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire FAC comp fail flag (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  flyByWireSECSwitch: new Offset({
    value: 0x7BA,
    name: 'flyByWireSECSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire SEC switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  flyByWireSECCompFailFlag: new Offset({
    value: 0x7BB,
    name: 'flyByWireSECCompFailFlag',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire SEC comp fail flag (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  apMasterSwitch: new Offset({
    value: 0x7BC,
    name: 'apMasterSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP master switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apWingLevel: new Offset({
    value: 0x7C0,
    name: 'apWingLevel',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP wing level',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apNav1Hold: new Offset({
    value: 0x7C4,
    name: 'apNav1Hold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP NAV1 hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apHeadingHold: new Offset({
    value: 0x7C8,
    name: 'apHeadingHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP heading hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apHeadingValue: new Offset({
    value: 0x7CC,
    name: 'apHeadingValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP heading value - degrees',
    convert: 'Math.round(({VAL} * 360) / 65536)',
    type: Type.UInt16,
    permission: 'rw',
  }),
  apAltitudeHold: new Offset({
    value: 0x7D0,
    name: 'apAltitudeHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP altitude hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAltitudeValue: new Offset({
    value: 0x7D4,
    name: 'apAltitudeValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP altitude value - ft',
    convert: '+({VAL} * 3.28084 / 65536).toFixed(2)',
    type: Type.Int32,
    permission: 'rw',
  }),
  apAttitudeHold: new Offset({
    value: 0x7D8,
    name: 'apAttitudeHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP attitude hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAsHold: new Offset({
    value: 0x7DC,
    name: 'apAsHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP airspeed hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAsValue: new Offset({
    value: 0x7E2,
    name: 'apAsValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP airspeed value - kt',
    type: Type.UInt16,
    permission: 'rw',
  }),
  apMachHold: new Offset({
    value: 0x7E4,
    name: 'apMachHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP mach hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apMachValue: new Offset({
    value: 0x7E8,
    name: 'apMachValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP mach value - mach',
    convert: '{VAL} / 65536',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apVsHold: new Offset({
    value: 0x7EC,
    name: 'apVsHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP vertical speed hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apVsValue: new Offset({
    value: 0x7F2,
    name: 'apVsValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP vertical speed value - ft/min',
    type: Type.Int16,
    permission: 'rw',
  }),
  apRPMN1Hold: new Offset({
    value: 0x7F4,
    name: 'apRPMN1Hold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP RPM/N1 hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apRPMN1Value: new Offset({
    value: 0x7FA,
    name: 'apRPMN1Value',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP RPM/N1 value - percent',
    convert: 'Math.round({VAL} * 100 / 16384)',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apGlideSlopeHold: new Offset({
    value: 0x7FC,
    name: 'apGlideSlopeHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP GlideSlope hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apApproachHold: new Offset({
    value: 0x800,
    name: 'apApproachHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP Approach hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apBackCourseHold: new Offset({
    value: 0x804,
    name: 'apBackCourseHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP back course hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apYawDamperHold: new Offset({
    value: 0x808,
    name: 'apYawDamperHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP yaw damper hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apTOGAAutoThrottle: new Offset({
    value: 0x80C,
    name: 'apTOGAAutoThrottle',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP TO/GA throttle hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAutoThrottleArm: new Offset({
    value: 0x810,
    name: 'apAutoThrottleArm',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP autothrottle arm',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apMaxBank: new Offset({
    value: 0x2E04,
    name: 'apMaxBank',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP max bank - degrees',
    type: Type.UInt32,
    permission: 'r',
  }),
  apMaxSpeed: new Offset({
    value: 0x2F40,
    name: 'apMaxSpeed',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP max speed',
    type: Type.Double,
    permission: 'r',
  }),
  apCruiseSpeed: new Offset({
    value: 0x2F48,
    name: 'apCruiseSpeed',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP cruise speed',
    type: Type.Double,
    permission: 'r',
  }),
};
