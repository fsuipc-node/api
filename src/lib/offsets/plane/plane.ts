import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const plane: OffsetList = {
  vs0: new Offset({
    value: 0x538,
    name: 'vs0',
    category: OffsetCategory.PLANE,
    description: 'Stall speed full flaps - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vs1: new Offset({
    value: 0x540,
    name: 'vs1',
    category: OffsetCategory.PLANE,
    description: 'Stall speed clean - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vc: new Offset({
    value: 0x548,
    name: 'vc',
    category: OffsetCategory.PLANE,
    description: 'Cruise speed - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vmd: new Offset({
    value: 0x550,
    name: 'vmd',
    category: OffsetCategory.PLANE,
    description: 'Minimum drag speed - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  smokeSystemControl: new Offset({
    value: 0x5D8,
    name: 'smokeSystemControl',
    category: OffsetCategory.PLANE,
    convert: '!!{VAL}',
    description: 'smoke system control on/off',
    type: Type.UInt16,
    permission: 'rw',
  }),
  retractableGear: new Offset({
    value: 0x60C,
    name: 'retractableGear',
    category: OffsetCategory.PLANE,
    description: 'retractable gear (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  retractableLeftFloatExtension: new Offset({
    value: 0x614,
    name: 'retractableLeftFloatExtension',
    category: OffsetCategory.PLANE,
    description: 'retractable left float extension - percent',
    convert: 'Math.floor({VAL} / 16384 * 100)',
    type: Type.UInt16,
    permission: 'r',
  }),
  retractableRightFloatExtension: new Offset({
    value: 0x616,
    name: 'retractableRightFloatExtension',
    category: OffsetCategory.PLANE,
    description: 'retractable right float extension - percent',
    convert: 'Math.floor({VAL} / 16384 * 100)',
    type: Type.UInt16,
    permission: 'r',
  }),
  hasFlaps: new Offset({
    value: 0x778,
    name: 'hasFlaps',
    category: OffsetCategory.PLANE,
    description: 'has flaps',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasStallHorn: new Offset({
    value: 0x77C,
    name: 'hasStallHorn',
    category: OffsetCategory.PLANE,
    description: 'has stall horn',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasMixtureEngine: new Offset({
    value: 0x780,
    name: 'hasMixtureEngine',
    category: OffsetCategory.PLANE,
    description: 'has mixture engine',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.PLANE,
    description: 'has carburator heater',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasSpoilers: new Offset({
    value: 0x78C,
    name: 'hasSpoilers',
    category: OffsetCategory.PLANE,
    description: 'has spoilers',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  isTailDragger: new Offset({
    value: 0x790,
    name: 'isTailDragger',
    category: OffsetCategory.PLANE,
    description: 'has spoilers',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasStrobes: new Offset({
    value: 0x794,
    name: 'hasStrobes',
    category: OffsetCategory.PLANE,
    description: 'has strobes',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasToeBrakes: new Offset({
    value: 0x79C,
    name: 'hasToeBrakes',
    category: OffsetCategory.PLANE,
    description: 'has toe brakes',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  maxMach: new Offset({
    value: 0xB0C,
    name: 'maxMach',
    category: OffsetCategory.PLANE,
    description: 'mach maximum operating - not FS2004',
    convert: '+({VAL} / 20438).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelSelectorsCount: new Offset({
    value: 0x123E,
    name: 'fuelSelectorsCount',
    category: OffsetCategory.PLANE,
    description: 'number of fuel selectors for the aircraft',
    type: Type.Byte,
    permission: 'r',
  }),
  aircraftEmptyWeight: new Offset({
    value: 0x1330,
    name: 'aircraftEmptyWeight',
    category: OffsetCategory.PLANE,
    description: 'aicraft empty weight',
    convert: '+({VAL} / 256).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
  aircraftMaxGrossWeight: new Offset({
    value: 0x1334,
    name: 'aircraftMaxGrossWeight',
    category: OffsetCategory.PLANE,
    description: 'aicraft max gross weight',
    convert: '+({VAL} / 256).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
};
