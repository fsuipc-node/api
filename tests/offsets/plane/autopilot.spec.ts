import { autopilot as offsets } from '@offsets/plane/autopilot';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/autopilot', () => {
  const offsetsTestCases = [
    { name: 'autoPilotAvailable', value: 0, expectedResult: false },
    { name: 'autoPilotAvailable', value: 1, expectedResult: true },
    { name: 'flyByWireELACSwitch', value: 0, expectedResult: false },
    { name: 'flyByWireELACSwitch', value: 1, expectedResult: true },
    { name: 'flyByWireELACCompFailFlag', value: 0, expectedResult: false },
    { name: 'flyByWireELACCompFailFlag', value: 1, expectedResult: true },
    { name: 'flyByWireFACSwitch', value: 0, expectedResult: false },
    { name: 'flyByWireFACSwitch', value: 1, expectedResult: true },
    { name: 'flyByWireFACCompFailFlag', value: 0, expectedResult: false },
    { name: 'flyByWireFACCompFailFlag', value: 1, expectedResult: true },
    { name: 'flyByWireSECSwitch', value: 0, expectedResult: false },
    { name: 'flyByWireSECSwitch', value: 1, expectedResult: true },
    { name: 'flyByWireSECCompFailFlag', value: 0, expectedResult: false },
    { name: 'flyByWireSECCompFailFlag', value: 1, expectedResult: true },
    { name: 'apMasterSwitch', value: 0, expectedResult: false },
    { name: 'apMasterSwitch', value: 1, expectedResult: true },
    { name: 'apWingLevel', value: 0, expectedResult: false },
    { name: 'apWingLevel', value: 1, expectedResult: true },
    { name: 'apNav1Hold', value: 0, expectedResult: false },
    { name: 'apNav1Hold', value: 1, expectedResult: true },
    { name: 'apHeadingHold', value: 0, expectedResult: false },
    { name: 'apHeadingHold', value: 1, expectedResult: true },
    { name: 'apHeadingValue', value: 32768, expectedResult: 180 },
    { name: 'apAltitudeHold', value: 0, expectedResult: false },
    { name: 'apAltitudeHold', value: 1, expectedResult: true },
    { name: 'apAltitudeValue', value: 19975000, expectedResult: 999.98 },
    { name: 'apAsHold', value: 0, expectedResult: false },
    { name: 'apAsHold', value: 1, expectedResult: true },
    { name: 'apMachHold', value: 0, expectedResult: false },
    { name: 'apMachHold', value: 1, expectedResult: true },
    { name: 'apMachValue', value: 229376, expectedResult: 3.5 },
    { name: 'apVsHold', value: 0, expectedResult: false },
    { name: 'apVsHold', value: 1, expectedResult: true },
    { name: 'apRPMN1Hold', value: 0, expectedResult: false },
    { name: 'apRPMN1Hold', value: 1, expectedResult: true },
    { name: 'apRPMN1Value', value: 9175, expectedResult: 56 },
    { name: 'apGlideSlopeHold', value: 0, expectedResult: false },
    { name: 'apGlideSlopeHold', value: 1, expectedResult: true },
    { name: 'apApproachHold', value: 0, expectedResult: false },
    { name: 'apApproachHold', value: 1, expectedResult: true },
    { name: 'apBackCourseHold', value: 0, expectedResult: false },
    { name: 'apBackCourseHold', value: 1, expectedResult: true },
    { name: 'apYawDamperHold', value: 0, expectedResult: false },
    { name: 'apYawDamperHold', value: 1, expectedResult: true },
    { name: 'apTOGAAutoThrottle', value: 0, expectedResult: false },
    { name: 'apTOGAAutoThrottle', value: 1, expectedResult: true },
    { name: 'apAutoThrottleArm', value: 0, expectedResult: false },
    { name: 'apAutoThrottleArm', value: 1, expectedResult: true },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
