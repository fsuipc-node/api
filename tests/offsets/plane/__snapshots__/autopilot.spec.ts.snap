// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/autopilot offsets list should have required properties 1`] = `
Object {
  "apAltitudeHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP altitude hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apAltitudeHold",
    "permission": "rw",
    "type": 6,
    "value": 2000,
  },
  "apAltitudeValue": Offset {
    "category": "autopilot",
    "convert": "+({VAL} * 3.28084 / 65536).toFixed(2)",
    "description": "AP altitude value - ft",
    "length": undefined,
    "mapping": undefined,
    "name": "apAltitudeValue",
    "permission": "rw",
    "type": 3,
    "value": 2004,
  },
  "apApproachHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP Approach hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apApproachHold",
    "permission": "rw",
    "type": 6,
    "value": 2048,
  },
  "apAsHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP airspeed hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apAsHold",
    "permission": "rw",
    "type": 6,
    "value": 2012,
  },
  "apAsValue": Offset {
    "category": "autopilot",
    "convert": undefined,
    "description": "AP airspeed value - kt",
    "length": undefined,
    "mapping": undefined,
    "name": "apAsValue",
    "permission": "rw",
    "type": 5,
    "value": 2018,
  },
  "apAttitudeHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP attitude hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apAttitudeHold",
    "permission": "rw",
    "type": 6,
    "value": 2008,
  },
  "apAutoThrottleArm": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP autothrottle arm",
    "length": undefined,
    "mapping": undefined,
    "name": "apAutoThrottleArm",
    "permission": "rw",
    "type": 6,
    "value": 2064,
  },
  "apBackCourseHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP back course hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apBackCourseHold",
    "permission": "rw",
    "type": 6,
    "value": 2052,
  },
  "apCruiseSpeed": Offset {
    "category": "autopilot",
    "convert": undefined,
    "description": "AP cruise speed",
    "length": undefined,
    "mapping": undefined,
    "name": "apCruiseSpeed",
    "permission": "r",
    "type": 8,
    "value": 12104,
  },
  "apGlideSlopeHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP GlideSlope hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apGlideSlopeHold",
    "permission": "rw",
    "type": 6,
    "value": 2044,
  },
  "apHeadingHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP heading hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apHeadingHold",
    "permission": "rw",
    "type": 6,
    "value": 1992,
  },
  "apHeadingValue": Offset {
    "category": "autopilot",
    "convert": "Math.round(({VAL} * 360) / 65536)",
    "description": "AP heading value - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "apHeadingValue",
    "permission": "rw",
    "type": 5,
    "value": 1996,
  },
  "apMachHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP mach hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apMachHold",
    "permission": "rw",
    "type": 6,
    "value": 2020,
  },
  "apMachValue": Offset {
    "category": "autopilot",
    "convert": "{VAL} / 65536",
    "description": "AP mach value - mach",
    "length": undefined,
    "mapping": undefined,
    "name": "apMachValue",
    "permission": "rw",
    "type": 6,
    "value": 2024,
  },
  "apMasterSwitch": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP master switch",
    "length": undefined,
    "mapping": undefined,
    "name": "apMasterSwitch",
    "permission": "rw",
    "type": 6,
    "value": 1980,
  },
  "apMaxBank": Offset {
    "category": "autopilot",
    "convert": undefined,
    "description": "AP max bank - degrees",
    "length": undefined,
    "mapping": undefined,
    "name": "apMaxBank",
    "permission": "r",
    "type": 6,
    "value": 11780,
  },
  "apMaxSpeed": Offset {
    "category": "autopilot",
    "convert": undefined,
    "description": "AP max speed",
    "length": undefined,
    "mapping": undefined,
    "name": "apMaxSpeed",
    "permission": "r",
    "type": 8,
    "value": 12096,
  },
  "apNav1Hold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP NAV1 hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apNav1Hold",
    "permission": "rw",
    "type": 6,
    "value": 1988,
  },
  "apRPMN1Hold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP RPM/N1 hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apRPMN1Hold",
    "permission": "rw",
    "type": 6,
    "value": 2036,
  },
  "apRPMN1Value": Offset {
    "category": "autopilot",
    "convert": "Math.round({VAL} * 100 / 16384)",
    "description": "AP RPM/N1 value - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "apRPMN1Value",
    "permission": "rw",
    "type": 6,
    "value": 2042,
  },
  "apTOGAAutoThrottle": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP TO/GA throttle hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apTOGAAutoThrottle",
    "permission": "rw",
    "type": 6,
    "value": 2060,
  },
  "apVsHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP vertical speed hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apVsHold",
    "permission": "rw",
    "type": 6,
    "value": 2028,
  },
  "apVsValue": Offset {
    "category": "autopilot",
    "convert": undefined,
    "description": "AP vertical speed value - ft/min",
    "length": undefined,
    "mapping": undefined,
    "name": "apVsValue",
    "permission": "rw",
    "type": 2,
    "value": 2034,
  },
  "apWingLevel": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP wing level",
    "length": undefined,
    "mapping": undefined,
    "name": "apWingLevel",
    "permission": "rw",
    "type": 6,
    "value": 1984,
  },
  "apYawDamperHold": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "AP yaw damper hold",
    "length": undefined,
    "mapping": undefined,
    "name": "apYawDamperHold",
    "permission": "rw",
    "type": 6,
    "value": 2056,
  },
  "autoPilotAvailable": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "autopilot available",
    "length": undefined,
    "mapping": undefined,
    "name": "autoPilotAvailable",
    "permission": "r",
    "type": 6,
    "value": 1892,
  },
  "flyByWireELACCompFailFlag": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "fly by wire ELAC comp fail flag (read only)",
    "length": undefined,
    "mapping": undefined,
    "name": "flyByWireELACCompFailFlag",
    "permission": "r",
    "type": 0,
    "value": 1975,
  },
  "flyByWireELACSwitch": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "fly by wire ELAC switch",
    "length": undefined,
    "mapping": undefined,
    "name": "flyByWireELACSwitch",
    "permission": "rw",
    "type": 0,
    "value": 1974,
  },
  "flyByWireFACCompFailFlag": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "fly by wire FAC comp fail flag (read only)",
    "length": undefined,
    "mapping": undefined,
    "name": "flyByWireFACCompFailFlag",
    "permission": "r",
    "type": 0,
    "value": 1977,
  },
  "flyByWireFACSwitch": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "fly by wire FAC switch",
    "length": undefined,
    "mapping": undefined,
    "name": "flyByWireFACSwitch",
    "permission": "rw",
    "type": 0,
    "value": 1976,
  },
  "flyByWireSECCompFailFlag": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "fly by wire SEC comp fail flag (read only)",
    "length": undefined,
    "mapping": undefined,
    "name": "flyByWireSECCompFailFlag",
    "permission": "r",
    "type": 0,
    "value": 1979,
  },
  "flyByWireSECSwitch": Offset {
    "category": "autopilot",
    "convert": "!!{VAL}",
    "description": "fly by wire SEC switch",
    "length": undefined,
    "mapping": undefined,
    "name": "flyByWireSECSwitch",
    "permission": "rw",
    "type": 0,
    "value": 1978,
  },
}
`;
